### ISSUE TITLE
update cFE unit tests

### ISSUE THREAD

**Issue by skliper** [nasa/cFE#65]
The version of the unit tests currently included with cFE 6.4.1 were developed and tested against cFE 6.3.2.  Due to cFE changes, all of the tests no longer pass.

Also, the tests have only been tested with the Linux OSAL/pc-linux PSP and ARINC653 OSAL/OrionSCP PSP.

JSC has updated the tests to work fully with cFE 6.4.1.  JSC is also updating the tests to build and run on VxWorks 6.7 (on the SPARC LEON3 processor).

These updates are being tracked in the JSC subversion repo and need to be pushed into a proper git branch and further work continued from there.

**Comment by skliper** [comment]
_Imported from trac issue 34.  Created by sduran on 2015-02-26T14:12:26, last modified: 2019-07-08T14:16:16_

**Comment by skliper** [comment]
_Trac comment by rmcclune on 2015-03-16 08:38:54:_

cFE branch "trac-34-update-cFE-unit-tests" created.  The unit tests are updated to cover cFE version 6.4.1.  Also, tests were modified and new tests added to improve the branch path and line coverage.

**Comment by skliper** [comment]
_Trac comment by gdecaruf on 2015-06-04 18:26:03:_

Update cFE for Class A & Associated unit tests
 
Expanding Scope of this ticket beyond unit tests to include cFE source code updates to all cFE files necessary to pursue a Class A safety-critical classification of this software for use by Orion and potentially others. The CFS AES project performed static code analysis utilizing the Understand tool for all MISRA rules, ran static cppcheck tool for potential safety concerns, ran the existing unit tests on a vxworks/LEON3 target with WindRiver Coverage analysis enabled to determine unit test code coverage, and held a general code inspection process. Specific changes as a result will be as follows:
 
-Updating all cFE files to comply with "most" MISRA rules and cppcheck static analysis errors
-Updating all cFE files to remove compiler warnings
-Expanding existing unit tests to provide full API-level test coverage based on exposed uncovered code
-Adding a full ADDITIONAL unit test suite using UTAssert framework to provide full code and BRANCH level coverage of the cFE source code.

**Comment by skliper** [comment]
_Trac comment by gdecaruf on 2015-06-09 19:25:10:_

'''Revisions to the CFE ES Services following Static Code Analysis:'''

commit: [changeset:d81a86c]

commit: [changeset:3a07d9f]

commit: [changeset:ad45ea7]

commit: [changeset:9c98e8f]

commit: [changeset:a759392]

commit: [changeset:a88fc92]

commit: [changeset:22facb6]

commit: [changeset:a37d2be]

 * Changed structure parameters from uint8 to char to satisfy Misra rule 6.1: All strings must have type char.  Changes in following structures:
   * cfe_es_apps.h - CFE_ES_AppStartParams_t, CFE_ES_MainTaskInfo_t,
   * cfe_es.h - CFE_ES_AppInfo_t, CFE_ES_CDSRegDumpRec_t
   * cfe_es_msg.h - CFE_ES_ShellCmd_t, CFE_ES_QueryAllCmd_t, CFE_ES_QueryAllTasksCmd_t, CFE_ES_StartAppCmd_t, CFE_ES_AppNameCmd_t, CFE_ES_TlmPoolStatsCmd_t, CFE_ES_ShellPacket_t
 * Changed prototype arguments for functions expecting string from uint8 to char in following functions:
   * CFE_ES_Main
   * CFE_ES_StartApplications
 * Made corresponding function call type cast changes for above functions in source code and unit test code (including changes in PSP - see !https://babelfish.arc.nasa.gov/trac/cfs_psp/ticket/10)
 * Corrected proper type casting to (void *) for PSP functions calls (CFE_PSP_MemSet, CFE_PSP_MemCpy, CFE_PSP_GetCFETextSegmentInfo, CFE_PSP_GetResetArea)
 * Corrected proper type casting to (void *) for OS function calls (OS_write, OS_read)
 * Replaced base type int to uint32 or int32 in local scopes in cfe_es_api.c
 * Deleted structure CFE_ES_DriverRegistryPacket_t in cfe_es_devsvr.h as it was dead code / unused.  The parameter CFE_ES_DeviceDriverRegistry was defined in this structure and as a global parameter, causing MISRA issue of redefinition in same scope.
 * Corrected types in CFE_ES_ObjectTable_t for ObjectPtr and StackPtr to use proper types. Parameters currently unused.
 * Corrected type of VoidPtr to osal_task_entry, to prevent compiler warning.
 * Replaced continue with conditional logic
 * Changed local parameters storing strlen return to size_t

**Comment by skliper** [comment]
_Trac comment by gdecaruf on 2015-06-09 21:10:35:_

'''Revision of cFE EVS Services following static code analysis:'''

commit: [changeset:e2fddf8]

 * Changed structure parameters from uint8 to char to satisfy Misra rule  6.1: All strings must have type char.  Changes following parameters in cfe_evs_msgs.h
   * !LogFilename
   * AppDataFilename
   * AppName
 * Corrected proper type casting to (void *) for PSP functions calls  (CFE_PSP_MemSet, CFE_PSP_MemCpy,   CFE_PSP_GetResetArea)
 * Corrected proper type casting to (void *) for OS function calls (OS_write)

**Comment by skliper** [comment]
_Trac comment by gdecaruf on 2015-06-11 17:14:09:_

'''Revision of cFE FS Services following static code analysis:'''

commit: [changeset:65695e5]

commit: [changeset:2d2d567]

 * Added proper casting to void * for PSP and OSAL function calls (CFE_PSP_MemCpy, CFE_PSP_MemSet, OS_write, OS_read)
 * Replaced call to memcpy to CFE_PSP_MemCpy
 * Changed basic int type to uint32 or int32 depending on the usage
 * Changed type of parameters storing size in bytes from strlen to size_t

**Comment by skliper** [comment]
_Trac comment by gdecaruf on 2015-06-11 19:08:05:_

'''Revision of cFE SB Services following static code analysis:'''

commit: [changeset:aab02be]

commit: [changeset:59640e0]

 * Replaced all continue calls with if/else logic.
 * Replaced a continue with break to fix a logic error resulting in infinite loop.
 * Added proper casting to void * for PSP and OSAL function calls (CFE_PSP_MemCpy, CFE_PSP_MemSet, OS_write)

**Comment by skliper** [comment]
_Trac comment by gdecaruf on 2015-06-11 20:25:49:_

'''Revision of cFE TBL Services following static code analysis:'''

commit: [changeset:1247881]

 * Replaced modification of index to break loop with break statement
 * Changed type of local parameters storing return from strlen as size_t
 * Added proper casting to void * for PSP and OSAL function calls (CFE_PSP_MemCpy, CFE_PSP_MemSet, OS_write, OS_read)

**Comment by skliper** [comment]
_Trac comment by gdecaruf on 2015-06-11 20:54:07:_

'''Revision of cFE TIME Services based on static code analysis:'''

commit: [changeset:d81a86c]

 * Added proper casting to void * of first argument in CFE_PSP_MemSet

commit: [changeset:976ef849]

 * Added empty Else statement after if / else if.

**Comment by skliper** [comment]
_Trac comment by gdecaruf on 2015-07-02 14:34:19:_

'''Revision following code analysis:  Change to U / UL for constants'''

commit: [changeset:f50e986e]

commit: [changeset:461d6e38]

commit: [changeset:1c36e160]

commit: [changeset:afb9ed0b]

commit: [changeset:2c23ab81]

commit: [changeset:fee2b856]

 * Added appropriate type cast U/L for HEX and large value constants
 * Uniform change for case of HEX values to lowercase with U/L as uppercase
 * Added type casting to (int32) for error codes which must be negative through overflow

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-07-06 21:47:48:_

I have taken a look at some of these changesets and I'm somewhat concerned that this is going far beyond unit test code.  Some of the modifications within these changesets directly duplicate other changes that have already been made and already accepted into development (case in point - changing the definitions in {{{cfe_error.h}}}).  These are likely to cause conflicts when we try to merge this back in again.

My feeling is that each of these individual commits listed here should be a separate trac ticket.  We should be merging this stuff back into development before it diverges too far from the mainline.

**Comment by skliper** [comment]
_Trac comment by sduran on 2015-07-07 14:21:45:_

I share the same concern.  That was not the original intent of these branches.  However, the code to be tested and the unit test code must match and they both live in the same repo.  So changes addressing fixes to issues found during unit testing, code reviews, and standards violations have also been made in the same branch.  Also, we need stable code to develop the unit tests against.  The development branch is a constant moving target now.  Once the test development is complete, then whom ever makes changes to the code can keep the unit test up the date too, but the unit tests were not there yet when you all jumped in and stated making changes.  It is unfortunate that others are making the same changes.  We will have to undergo the pain of merging this in at some point into the development branch after we have the completed the class A cert in a couple of months.

**Comment by skliper** [comment]
_Trac comment by jphickey on 2015-07-07 15:47:39:_

My opinion here is that the class A stuff deserves its own development branch.  

We already discussed the idea that in addition to "master" and the main "development" branch we could have additional development branches targeting specific audiences.  I have already made such a branch for the electronic data sheet version.

Small, to-the-point change sets tend to be easy to merge/move around and as such can be merged to __both__ normal "development" and the class-a development branch (TBD).

I think this approach helps everyone as it gets the useful stuff from all this class A work into the main "development" tree so we all can benefit from it, as well as getting it built and executed more frequently by using the Bamboo system.  Not only that, it saves you the effort of re-doing changes that were already done months ago (see [changeset:01adb82]/#53 - committed 02/05/15), and saves us from future merge conflict pain.  Seems like a win-win.  

Other examples of already-done fixes that address compiler warnings & casting: #51 #54 #55 #57 #58 #64.  These were all intentionally done as isolated change sets such that they __could__ be taken by themselves into more conservative branches for those that do not want the rest of the new features.

To further illustrate the problem here: the version this is based upon is still 6.4.1 even though we have 6.4.2 tagged and released that fixes a known serious start up race condition and also fixes a segmentation fault in the unit test.  What is the plan to bring this up to 6.4.2?

**Comment by skliper** [comment]
_Trac comment by gdecaruf on 2015-08-11 15:49:36:_

Change following code change review:

commit: [changeset:75ada6f8]

Fixed missed case correction for Hex Value.  All Hex values have been changed to lower-case for consistency.  The type indicator is in Upper case (L, UL, U)

Example:  0xabcdef12UL

**Comment by skliper** [comment]
_Trac comment by gdecaruf on 2015-08-11 16:12:03:_

Changes following code change review:

commit: [changeset:0149d12a]

Removed unnecessary type casts for small values and outputs.

**Comment by skliper** [comment]
_Trac comment by gdecaruf on 2015-08-11 17:08:09:_

Change following code change review:

commit: [changeset:58be7d34]

Logic correction for filename length.  Initial correction planned was to change DirMarkIdx to ssize_t.  However ssize_t is POSIX specific and not portable.  Instead, a correction factor is added to insure that the original logic is maintained.  The original logic had DirMarkIdx initialized to -1, before it's type was corrected in: [changeset:2d2d5671].

**Comment by skliper** [comment]
_Trac comment by glimes on 2016-04-19 22:46:27:_

This ticket got "parted out" and we believe that
we have dealt with all of the parts. Closing.

**Comment by skliper** [comment]
_Trac comment by jhageman on 2019-03-05 14:57:55:_

Milestone renamed

**Comment by skliper** [comment]
_Trac comment by jhageman on 2019-07-08 14:16:16:_

Moved work to techdev-JSC-Class-A_and_unit-tests branch and closed trac branch since ticket is closed.
